test/assert_options.go

```diff
@@ -29,14 +15,58 @@ import (
 type TestingOption func(*testingConfig) error
 
 type testingConfig struct {
-	backends             []backend.ID
-	curves               []ecc.ID
-	witnessSerialization bool
-	solverOpts           []solver.Option
-	proverOpts           []backend.ProverOption
-	compileOpts          []frontend.CompileOption
-	fuzzing              bool
-	solidity             bool
+	profile
+	solverOpts  []solver.Option
+	proverOpts  []backend.ProverOption
+	compileOpts []frontend.CompileOption
+
+	validAssignments   []frontend.Circuit
+	invalidAssignments []frontend.Circuit
+}
+
+// default options
+func (assert *Assert) options(opts ...TestingOption) testingConfig {
+	var opt testingConfig
+
+	// default options;
+	// go test -short 					--> testEngineOnly
+	// go test 							--> constraintOnlyProfile
+	// go test -tags=prover_checks 		--> proverOnlyProfile
+	// go test -tags=release_checks 	--> releaseProfile
+
+	if releaseTestFlag {
+		opt.profile = releaseChecks
+	} else if proverTestFlag {
+		opt.profile = proverChecks
+	} else if testing.Short() {
+		opt.profile = testEngineChecks
+	} else {
+		opt.profile = constraintSolverChecks
+	}
+
+	// apply user provided options.
+	for _, option := range opts {
+		err := option(&opt)
+		assert.NoError(err, "parsing TestingOption")
+	}
+
+	return opt
+}
+
+// WithValidAssignment is a testing option which adds a valid assignment
+func WithValidAssignment(validAssignment frontend.Circuit) TestingOption {
```

Comment:
What about having a variation of this option (in addition to this one) called `WithValidAssignmentMaker` (etc instead of `Maker`) which constructs a new witness every time is called? Sometimes when we have some complex witness types (a la `Element` in non-native package) then we change some internal fields here and there to keep track of the usage inside circuit (for example, if something has been range-checked).

This may lead to failures in determinism tests. Even though the circuits can be fixed (and I usually have, by implementing cleanup or storing some stuff in internal caches of the non-native API), it may be quite inconvenient.